_posts/2021-03-10-dask_memory_usage.md

```diff
@@ -0,0 +1,261 @@
+---
+layout: post
+title: Measuring Dask memory usage with dask-memusage
+author: <a href="https://pythonspeed.com">Itamar Turner-Trauring</a>
+tags: [memory, dask, distributed, profiling, ram]
+theme: twitter
+---
+
+{% include JB/setup %}
+
+When thinking about the bottlenecks in large-scale computation, we often focus on CPU: we want to use more CPU cores in order to get faster results.
+But memory can also be a bottleneck, leading to out-of-memory errors, or forcing under-utilization of available CPUs.
+
+Unfortunately, memory usage can often be difficult to measure and understand.
+So in this article I'm going to introduce [dask-memusage](https://github.com/itamarst/dask-memusage/), a tool that lets your see peak memory usage of the tasks in your Dask execution graph.
+This can help you determine necessary hardware resources, and point at code that might need its memory usage reduced.
+
+Once you've discovered where the memory bottleneck is, you might be able to ensure maximum hardware utilization by tweaking your configuration, especially if you're running on cloud VMs.
+Or, you can try to reduce memory usage, for example by using the [Fil memory profiler](https://pythonspeed.com/fil) to further pinpoint the exact lines of code responsible for high memory usage.
+
+## The problem: fixed processing chunks and a high memory/CPU ratio
```

Comment:
Something that is confusing about this post and makes it hard to read is that there are lots of different examples which aren't always clearly distinguished (maybe headings or subheadings would help). As a result, it's hard to keep them all straight, and remember what the takeaway message was from each one.

1. A real world example: image processing for spatial gene sequencing
    * Why it's introduced: To give people an idea of a real world use case
    * What the takeaway is: I don't know what you wanted the audience to take away from this, we never return to this example.
2. A toy example: squaring numbers
    * Why it's introduced: This isn't clearly stated.
    * What the takeaway is: This isn't clearly stated either. 
3. Another toy example: counting words
    * Why it's introduced: So we have a small, understandable example to practice on
    * What the takeaway is: Now we have learned how to use dask-memusage and Fil to identify problem areas with memory use

It's also worth trying to explain clearly how example 2 & 3 compare to each other, too.